src/tree/tasks.rs

```diff
@@ -0,0 +1,147 @@
+        //       of the bridged networks. However we don't have 100% guarantee that the canonical network
+        //       events will always arrive before the bridged network events.
+        //       So to maintain liveliness we reallign on every update.
+        reallign_trees(&mut identity_tree, &mut chain_state).await;
+    }
+
+    Err(WorldTreeError::LeafChannelClosed)
+}
+
+pub async fn handle_bridged_updates<M>(
+    identity_tree: Arc<RwLock<IdentityTree<MmapVec<Hash>>>>,
+    chain_state: Arc<RwLock<HashMap<u64, Hash>>>,
+    mut bridged_root_rx: Receiver<(u64, Hash)>,
+    mut cancel_rx: broadcast::Receiver<()>,
+) -> Result<(), WorldTreeError<M>>
+where
+    M: Middleware + 'static,
+{
+    loop {
+        let (chain_id, bridged_root) = tokio::select! {
+            res = bridged_root_rx.recv() => {
+                match res {
+                    Some((chain_id, bridged_root)) => (chain_id, bridged_root),
+                    None => break,
+                }
+            }
+            _ = cancel_rx.recv() => {
+                break
+            }
+        };
+
+        tracing::info!(?chain_id, root = ?bridged_root, "Bridged root received");
+
+        let mut identity_tree = identity_tree.write().await;
+        let mut chain_state = chain_state.write().await;
+
+        // Update chain state with the new root
+        chain_state.insert(chain_id, bridged_root);
+
+        reallign_trees(&mut identity_tree, &mut chain_state).await;
+    }
+
+    Err(WorldTreeError::BridgedRootChannelClosed)
+}
+
+/// Realligns all the observed chains.
+///
+/// This function figures out the root that has been seen across all observed networks
+/// And applies all the updates up to that root to the canonical tree.
+async fn reallign_trees(
+    identity_tree: &mut IdentityTree<MmapVec<Hash>>,
+    chain_state: &mut HashMap<u64, Hash>,
+) {
+    let mut chain_state_idxs = vec![];
+
+    for (chain_id, root) in chain_state.iter() {
+        let idx = identity_tree
+            .tree_updates
+            .iter()
+            .position(|(hash, _updates)| hash == root);
+
+        //
+        if let Some(idx) = idx {
+            chain_state_idxs.push(idx);
+        } else {
+            // This can occur in 2 cases:
+            // 1. The bridged network event arrived before the canonical event
+            //    this is very unlikely to happen in prod, but technically possible (RPC outage, etc.)
+            // 2. A bridged network is still at the cascading tree root
+            //
+            // in both cases, we cannot reallign the trees
+            tracing::warn!("Root {root:?} seen on chain {chain_id} not found in tree updates - cannot reallign");
+            return;
+        }
+    }
+
+    let Some(latest_common_root_idx) = chain_state_idxs.iter().min().copied()
+    else {
+        // If we don't find any common roots then there's nothing to reallign
+        return;
+    };
```

Comment:
IIUC, this logic is iterating through the `chain_state` and finding which tree update corresponds with the current root on each chain. Then it is pushing the corresponding index for that update to `chain_state_idxs`. 

Then you are getting the min `chain_state_idx` and applying updates to this root, however `let Some(latest_common_root_idx) = chain_state_idxs.iter().min().copied()` is getting the minimum index across all chains rather than the index representing the oldest common root across chains.

Since `reallign_trees` runs on both canonical updates and bridged updates, I believe the following scenarios are possible. Let me know if I am overlooking something.

**Scenario 1:**
- Canonical root update `0xa`
- Bridged root update missed
- Canonical root update `0xb`
  - returns early
- Bridged root update `0xb`
  - Min root is `0xa`, updates applied to `0xa` but updates should be applied to `0xb`, causing state to perpetually lag behind.

**Scenario 2:**
- Canonical root update `0xc` missed
- Bridged root update `0xc`
  - Panic when trying to apply updates due to:

```rust
  let idx_of_root = self
            .tree_updates
            .iter()
            .position(|(update_root, _update)| update_root == root)
            .expect("Tried applying updates to a non-existent root");
```

Reply:
I think in scenario 1 we'll actually proceed correctly like so:
1. We get the canonical root 0xa along with leaf updates
2. Bridged chain misses this update
3. Canonical update 0xb
4. Same goes to 0xb

Since we now see 0xb as observed by both chains it'll be detected as the latest common root. And we'll apply the updates up to it

Reply:
Scenario 2 is worse - we'll actually panic earlier in `append_updates`. We might consider having BlockScanner lag behind chain head. Or have a second instance of BlockScanner lagging behind trying to plug missing updates. I'm open to ideas though

Reply:
I can add a test for scenario 1

Reply:
And to put the logic in words:

On every update we save the known observed root on each chain.
In case of the canonical chain we also calculate the pending storage updates.

On every update we find the index of the root in each observed chain - and the smallest one. Since order of updates in `tree_updates` is maintained by the pre/post root chaining we know that the smallest index can be merged in to the cascading tree.